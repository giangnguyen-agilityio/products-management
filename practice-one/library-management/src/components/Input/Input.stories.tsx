import {type StoryFn, type Meta} from '@storybook/react'
import Input from './index'
import React from 'react'

export default {
  title: 'Example/Input',
  component: Input,
} as Meta<typeof Input>

const Template: StoryFn<typeof Input> = args => <Input {...args} />

export const Default = Template.bind({})
Default.args = {
  label: 'Username',
  classNameLabel: 'username',
  type: 'text',
  name: 'username',
  placeholder: 'Enter your username',
  className: 'username-input',
  id: 'username',
  errorMessage: 'Please enter your username',
}
