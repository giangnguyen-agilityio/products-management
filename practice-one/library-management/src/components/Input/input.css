.input-wrapper {
  display: flex;
  flex-direction: column;
  gap: 5px;
}

.label {
  text-transform: capitalize;
}

.input {
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid var(--secondary-color);
  border-radius: 5px;
  padding: 5px 8px;
  transition: outline-color 0.2s ease-in-out;
}

.input:hover {
  cursor: pointer;
}

.input:focus {
  outline-color: var(--primary-color);
}

.error-message {
  color: var(--danger-color);
  font-family: var(--secondary-font);
}
