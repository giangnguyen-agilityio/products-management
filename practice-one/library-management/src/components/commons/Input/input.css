.input-wrapper {
  display: flex;
  flex-direction: column;
  gap: 5px;
  margin: 15px 0;
}

.label {
  text-transform: inherit;
}

.input {
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 5px;
  transition: outline-color 0.2s ease-in-out;
  border: 1px solid var(--secondary-color);
  padding: 5px 8px;
  font-size: 16px;
  font-family: var(--primary-font);
}

.input:hover {
  cursor: pointer;
}

.input:focus {
  outline-color: var(--primary-color);
}

.error-message {
  color: var(--danger-color);
  font-family: var(--secondary-font);
}
