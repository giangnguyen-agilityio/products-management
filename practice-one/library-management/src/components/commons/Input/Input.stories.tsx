import {type StoryFn, type Meta} from '@storybook/react'
import Input from './index'
import React from 'react'

export default {
  title: 'Example/Input',
  component: Input,
} as Meta<typeof Input>

const Template: StoryFn<typeof Input> = args => <Input {...args} />

export const TextInput = Template.bind({})
TextInput.args = {
  label: 'Username:',
  classNameLabel: 'username',
  type: 'text',
  name: 'username',
  placeholder: 'Enter your username',
  className: 'username-input',
  id: 'username',
  errorMessage: 'Please enter your username',
}

export const DateTimeInput = Template.bind({})
DateTimeInput.args = {
  label: 'Date Time Local:',
  classNameLabel: 'datetime-local-label',
  type: 'datetime-local',
  className: 'datetime-local',
}

export const UploadImageInput = Template.bind({})
UploadImageInput.args = {
  label: 'Upload Your Image:',
  classNameLabel: 'upload-image-label',
  className: 'upload-file',
  type: 'file',
  accept: 'image/*',
}

export const NumberInput = Template.bind({})
NumberInput.args = {
  label: 'Enter your quantity:',
  classNameLabel: 'quantity-input-label',
  className: 'quantity-input',
  type: 'number',
  min: 0,
}
