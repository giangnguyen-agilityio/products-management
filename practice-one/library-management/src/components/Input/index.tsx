import React from 'react'

// Importing the CSS file for styling
import './input.css'

// Define the props for the Input component
interface InputProps {
  label?: string
  classNameLabel?: string
  type: string
  name: string
  placeholder?: string
  className?: string
  value: string
  id: string
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
  errorMessage?: string
}

const Input: React.FC<InputProps> = (props): JSX.Element => {
  const {
    classNameLabel,
    label,
    type,
    name,
    placeholder,
    className,
    value,
    id,
    onChange,
    errorMessage
  } = props
  return (
    <div className="input-wrapper">
      <label htmlFor={id} className={`label ${classNameLabel ?? ''}`}>
        {label}
      </label>
      <input
        type={type}
        name={name}
        placeholder={placeholder}
        value={value}
        className={`input ${className ?? ''}`}
        id={id}
        onChange={onChange}
      />
      {errorMessage != null && <p className="error-message">{errorMessage}</p>}
    </div>
  )
}

export default Input
