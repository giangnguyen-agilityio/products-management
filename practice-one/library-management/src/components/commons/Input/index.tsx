import React, {memo, type InputHTMLAttributes} from 'react'
import './input.css'

interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  name?: string
  label?: string
  className?: string
  classNameLabel?: string
  value?: string | number
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
  errorMessage?: string
  type?: string
  min?: number
  accept?: string
  placeholder?: string
}

const Input: React.FC<InputProps> = (props): JSX.Element => {
  const {
    name,
    label,
    className,
    classNameLabel,
    value,
    onChange,
    errorMessage,
    type,
    min,
    accept,
    placeholder,
  } = props
  return (
    <div className="input-wrapper">
      <label className={`label ${classNameLabel ?? ''}`}>{label}</label>
      <input
        className={`input ${className ?? ''}`}
        name={name}
        type={type ?? 'text'}
        value={value}
        onChange={onChange}
        min={min ?? 0}
        accept={accept}
        placeholder={placeholder}
      />
      {errorMessage != null && <p className="error-message">{errorMessage}</p>}
    </div>
  )
}

export default memo(Input)
