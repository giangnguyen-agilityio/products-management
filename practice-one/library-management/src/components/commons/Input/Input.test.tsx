import React, {InputHTMLAttributes} from 'react'
import {render, fireEvent} from '@testing-library/react'
import Input from '.'

interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  name?: string
  label?: string
  className?: string
  classNameLabel?: string
  value?: string | number
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
  errorMessage?: string
  type?: string
  min?: number
  accept?: string
  placeholder?: string
}

const renderInput = (props: InputProps) => {
  return render(<Input {...props} />)
}

describe('Input Component', () => {
  it('should render Input component with basic props', () => {
    const {container, getByPlaceholderText} = renderInput({
      label: 'Username',
      name: 'username',
      placeholder: 'Enter your username',
    })

    const inputElement = container.querySelector('input')
    const placeholderElement = getByPlaceholderText('Enter your username')

    expect(inputElement).toBeInTheDocument()
    expect(placeholderElement).toBeInTheDocument()
    expect(inputElement?.tagName).toBe('INPUT')

    expect(inputElement).toMatchSnapshot()
  })

  it('should render Input component with a custom className for the label', () => {
    const {container} = renderInput({
      label: 'Email',
      name: 'email',
      classNameLabel: 'custom-label',
    })

    const labelElement = container.querySelector('label')

    expect(labelElement).toBeInTheDocument()
    expect(labelElement?.classList).toContain('custom-label')

    expect(labelElement).toMatchSnapshot()
  })

  it('should render Input component with an error message', () => {
    const {getByText} = renderInput({
      label: 'Age',
      name: 'age',
      errorMessage: 'Age must be a number',
    })

    const errorMessageElement = getByText('Age must be a number')
    expect(errorMessageElement).toBeInTheDocument()

    expect(errorMessageElement).toMatchSnapshot()
  })

  it('should fire onChange event when input value changes', () => {
    const mockOnChange = jest.fn()
    const {container} = renderInput({
      label: 'Password',
      name: 'password',
      type: 'password',
      onChange: mockOnChange,
    })

    const inputElement = container.querySelector('input')
    fireEvent.change(inputElement!, {target: {value: 'newPassword'}})

    expect(mockOnChange).toHaveBeenCalledTimes(1)
    expect(inputElement?.type).toBe('password')
    expect(inputElement?.value).toBe('newPassword')

    expect(container).toMatchSnapshot()
  })

  it('should render Input component with type, min, and accept attributes', () => {
    const {container} = renderInput({
      label: 'Upload File',
      name: 'file',
      type: 'file',
      min: 1,
      accept: '.jpg,.png',
    })

    const inputElement = container.querySelector('input')

    expect(inputElement).toBeInTheDocument()
    expect(inputElement?.type).toBe('file')
    expect(inputElement?.min).toBe('1')
    expect(inputElement?.accept).toBe('.jpg,.png')

    expect(inputElement).toMatchSnapshot()
  })
})
