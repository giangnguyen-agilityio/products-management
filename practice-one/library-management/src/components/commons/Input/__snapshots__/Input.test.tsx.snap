// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component should fire onChange event when input value changes 1`] = `
<div>
  <div
    class="input-wrapper"
  >
    <label
      class="label "
    >
      Password
    </label>
    <input
      class="input "
      min="0"
      name="password"
      type="password"
      value=""
    />
  </div>
</div>
`;

exports[`Input Component should render Input component with a custom className for the label 1`] = `
<label
  class="label custom-label"
>
  Email
</label>
`;

exports[`Input Component should render Input component with an error message 1`] = `
<p
  class="error-message"
>
  Age must be a number
</p>
`;

exports[`Input Component should render Input component with basic props 1`] = `
<input
  class="input "
  min="0"
  name="username"
  placeholder="Enter your username"
  type="text"
  value=""
/>
`;

exports[`Input Component should render Input component with type, min, and accept attributes 1`] = `
<input
  accept=".jpg,.png"
  class="input "
  min="1"
  name="file"
  type="file"
  value=""
/>
`;
